keps/sig-cluster-lifecycle/0032-create-a-k8s-io-component-repo.md

```diff
@@ -0,0 +1,245 @@
+- Cloud providers can be moved out of core without having to depend on the core repository.
+  - Related issue: [https://github.com/kubernetes/kubernetes/issues/69585](https://github.com/kubernetes/kubernetes/issues/69585)
+- It's easier for _kubeadm_ to move out of the core repo when these component-related packages are in a "public" staging repository.
+
+### Non-goals
+
+- Graduate any ComponentConfig API versions (in this proposal).
+- Make this library toolkit a "generic" cloud-native component builder. Such a toolbox, if ever created, could instead consume these packages. In other words, this repository is solely focused on Kubernetes' components' needs.
+- Fixing _all the problems_ our components have right now, and expanding this beyond what's really necessary. Instead working incrementally, and starting to break out some basic stuff we _know_ every component must handle (e.g. configuration and flag parsing)
+
+### Related proposals / references
+
+- [Kubernetes Component Configuration](https://docs.google.com/document/d/1arP4T9Qkp2SovlJZ_y790sBeiWXDO6SG10pZ_UUU-Lc/edit) by [@mikedanese](https://github.com/mikedanese)
+- [Versioned Component Configuration Files](https://docs.google.com/document/d/1FdaEJUEh091qf5B98HM6_8MS764iXrxxigNIdwHYW9c/edit#) by [@mtaufen](https://github.com/mtaufen)
+- [Moving ComponentConfig API types to staging repos](https://github.com/kubernetes/community/blob/master/keps/sig-cluster-lifecycle/0014-20180707-componentconfig-api-types-to-staging.md) by [@luxas](https://github.com/luxas) and [@sttts](https://github.com/sttts)
+
+## Proposal
+
+This proposal contains three logical units of work. Each subsection is explained in more detail below.
+
+### Part 1: ComponentConfig
+
+#### Standardized encoding/decoding
+
+- Encoding/decoding helper methods in `k8s.io/component/config/serializer` that would be referenced in every scheme package
+- Warn or (if desired, error) on unknown fields
+  - This has the benefit that it makes it possible for the user to spot e.g. config typos. More high-level, this can be used for e.g. a `--validate-config` flag
+- Support both JSON and YAML
+- Support multiple YAML documents if needed
+
+#### Testing helper methods
+
+- Conversion / roundtrip testing
+- API group testing
+  - External types must have JSON tags
+  - Internal types must not have any JSON tags
+  - ...
+- Defaulting testing
+
+#### Generate OpenAPI specifications
+
+Provide a common way to generate OpenAPI specifications local to the component, so that external consumers can access it, and the component can expose it via e.g. a CLI flag or HTTPS endpoint.
+
+### Part 2: Command building / flag parsing
+
+#### Wrapper around cobra.Command
+
+See the `cmd/kubelet` code for how much extra setup a Kubernetes component needs to do for building commands and flag sets. This code can be refactored into a generic wrapper around _cobra_ for use with Kubernetes.
```

Comment:
it's not clear to me that most components actually *want* cobra -- it makes things way more complicated than straight-up pflag, and doesn't win a lot for things that look more like controllers (it's super-useful for CLIs, though).